### PR TITLE
ThemeDictionary + ThemeVariantScope

### PR DESCRIPTION
**Pull request by maxkatz6**
## What does the pull request do?

Up to date API changes:
Avalonia.Base:
```diff
+public sealed record ThemeVariant
+{
+    public ThemeVariant(object key, ThemeVariant? inheritVariant);
+    public static ThemeVariant Default { get; }
+    public static ThemeVariant Light { get; }
+    public static ThemeVariant Dark { get; }
+    public object Key { get; }
+    public ThemeVariant? InheritVariant { get; }
+}
public interface IResourceDictionary
{
+    IDictionary<ThemeVariant, IResourceProvider> ThemeDictionaries { get; }
}
public interface IResourceNode
{
-    bool TryGetResource(object key, out object? value);
+    bool TryGetResource(object key, ThemeVariant? theme, out object? value);
}
public class StyledElement
{
+    public static readonly StyledProperty<ThemeVariant> ActualThemeVariantProperty;
+    public static readonly StyledProperty<ThemeVariant?> RequestedThemeVariantProperty;
+    public ThemeVariant ActualThemeVariant { get; }
}
+public interface IGlobalThemeVariantProvider : IResourceHost
+{
+    ThemeVariant ActualThemeVariant { get; }
+    event EventHandler? ActualThemeVariantChanged;
+}
```
Avalonia.Controls:
```diff
+public class ThemeVariantScope : StyledElement
+{
+    public ThemeVariant? RequestedThemeVariant { get; set; }
+}
public class Application : IGlobalThemeVariantProvider
{
+    public ThemeVariant? RequestedThemeVariant { get; set; }
+    public ThemeVariant ActualThemeVariant { get; }
+    event EventHandler? ActualThemeVariantChanged
}
public class TopLevel : StyledElement
{
+    public ThemeVariant? RequestedThemeVariant { get; set; }
}
```

Avalonia.Theme.xxx
```diff
public class FluentTheme
{
-    FluentThemeMode Mode
}
-public enum FluentThemeMode { Light, Dark }
public class SimpleTheme
{
-    SimpleThemeMode Mode
}
-public enum SimpleThemeMode { Light, Dark }
```

P.S. IGlobalThemeVariantProvider can be avoided if we directly subscribe on Application.PropertyChanged from the TopLevel. As property inheriting doesn't work from the Application to the Window normally. cc @grokys 

### How to define a set of theme specific resources

In Avalonia, theme variant specific resources can be defined in the `ResourceDictionary` using the `ThemeDictionaries` dictionary. The `ThemeDictionaries` dictionary uses `ThemeVariant` as the key and a child `ResourceDictionary` as the value.

Typically, developers use `Light` or `Dark` as the key for the theme variants, which are recognized at compile time. Using `Default` as the key marks a specific theme dictionary as a fallback in case the theme variant or resource key is not found in other theme dictionaries.

In addition to the built-in values of `Light`, `Dark`, and `Default`, any object value can be used as a key (since it's wrapped in the `ThemeVariant(object key)` structure). `{x:Static}` markup extension can also be used here, if a developer wants to define multiple custom themes as static properties and reference them from the XAML code.

Example:
```xaml
<ResourceDictionary>
    <ResourceDictionary.ThemeDictionaries>
        <ResourceDictionary x:Key='Default'>
            <SolidColorBrush x:Key='BackgroundBrush'>White</SolidColorBrush>
            <SolidColorBrush x:Key='ForegroundBrush'>Black</SolidColorBrush>
        </ResourceDictionary>
        <ResourceDictionary x:Key='Dark'>
            <SolidColorBrush x:Key='BackgroundBrush'>Black</SolidColorBrush>
            <SolidColorBrush x:Key='ForegroundBrush'>White</SolidColorBrush>
        </ResourceDictionary>
    </ResourceDictionary.ThemeDictionaries>
</ResourceDictionary>
```
Note, in this PR for the Simple and Fluent theme I defined light-variant resources as `Default`. It's an exact opposite of UWP definitions, which define dark-variant resources as a `Default`.

Why? It's theme specific and we historically used light-variant as a default. Not to mention, most of modern OSes are light by default now. Custom third-party themes can still define it how they decide.

### Defining custom ThemeVariant:

Custom Theme Variant can be defined as:
```csharp
public static class Themes
{
    public static ThemeVariant Pink { get; } = new ThemeVariant(nameof(Pink))
    {
        InheritVariant = ThemeVariant.Light
    };
}
```
Then, when a resource lookup is performed, the theme dictionary defined with the `Pink` key is selected. If the resource is not found in that dictionary, the `Light` variant is used as a fallback.
It is important for developers to always set `InheritVariant`, as otherwise Avalonia will not be able to convert the application-level variant to the system-level variant (which is always either `Light` or `Dark`) in order to change the window frame colors.
However, when the system theme variant is used (i.e., no requested variant is set), it will always use the "Dark" or "Light" default variants. In the future, there may be a need for an API to convert system-`Light` to a custom application-`Pink` variant, for example.

### The theme variant lookup in this PR works as follows:

Each control has two properties: `RequestedThemeVariant` and `ActualThemeVariant`. `RequestedThemeVariant` is only accessible for `Application`, `TopLevel` (including `Window`), and `ThemeVariantScrope` controls. Potentially it can be added to the `Page` control to improve mobile support. We need this restriction because Avalonia has inheritable properties, while UWP does not. This would cause problems with font and background colors inherited from the parent which has different theme variant. While `ThemeVariantScrope` override most common inheritable properties.

When `Application.RequestedThemeVariant` is not set, the application will inherit the system theme variant. The same goes for Window and its child controls tree, which will inherit the `ActualThemeVariant` from the application. If `RequestedThemeVariant` is set on any supported control, the subtree will have an overridden `ActualThemeVariant`.

Setting `RequestedThemeVariant` to `Default` or `null` will reset actual theme and value will be inherited from the parent (or from the system in case of Application.RequestedThemeVariant).

It's worth noting that there is a design flaw in that a user can still go one level lower and call `control.SetValue(RequestedThemeVariant)` or `control.SetValue(ActualThemeVariant)` on any control, which will still work but will not override inherited controls automatically. Ideally, `ActualThemeVariant` should be a fully read-only property, but it also needs to be inheritable, which is not currently possible to combine in Avalonia.

### The resource lookup in this PR works as follows:

The `ThemeVariant` is picked up from the `StyledControl.ActualThemeVariant`, regardless of whether the extension was used on the control or attached with a style. Each `ResourceDictionary` looks up the resource in the following priority:

1. Checks if the resource is defined directly in the `ResourceDictionary`.
2. If a `ThemeVariant` is provided:
2.1. Uses the theme dictionary with the variant key, if defined.
2.2. Otherwise, uses `ThemeVariant.InheritVariant` recursively, if defined.
3. If resource wasn't found, uses the theme dictionary with the `Default` key, if defined.
4. Checks if the resource is defined in any merged resources.

Steps 2 and 3 are new and may add overhead to existing applications, but only if the `ResourceDictionary` has at least one theme dictionary set.

In the case of `StaticResource`, when an extension is used from the `ResourceDictionary`, it won't use any theme variant during the lookup process. Additionally, it is not able to understand if the extension was used from a specific theme dictionary, as our markup extension does not have this context information. However, it will still work if both resources are defined in the same dictionary or have been merged into a single dictionary. This limitation is similar to that of UWP, which also merges dictionaries.

## Checklist

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [x] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Breaking changes

Yes, SimpleThemeMode and FluentThemeMode removed. IResourceNode interface was changed.

https://user-images.githubusercontent.com/3163374/169640920-7fd91523-7a95-496e-979d-95934a790e50.mp4

### CODE REVIEW

**Comment by maxkatz6** [comment]
First benchmarks...

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.25115
Intel Core i5-9600K CPU 3.70GHz (Coffee Lake), 1 CPU, 6 logical and 6 physical cores
.NET SDK=6.0.300
  [Host]     : .NET 6.0.5 (6.0.522.21309), X64 RyuJIT
  DefaultJob : .NET 6.0.5 (6.0.522.21309), X64 RyuJIT

#### ControlsBenchmark
Master:
|         Method |        Mean |     Error |    StdDev |    Gen 0 |    Gen 1 | Allocated |
|--------------- |------------:|----------:|----------:|---------:|---------:|----------:|
| CreateCalendar | 19,261.4 us | 169.02 us | 158.10 us | 687.5000 | 281.2500 |  4,258 KB |
|   CreateButton |    136.6 us |   1.10 us |   0.97 us |   8.7891 |   1.2207 |     41 KB |
|  CreateTextBox |    845.3 us |  16.38 us |  15.32 us |  46.8750 |   7.8125 |    217 KB |

This PR:
|         Method |        Mean |     Error |    StdDev |    Gen 0 |    Gen 1 |   Gen 2 | Allocated |
|--------------- |------------:|----------:|----------:|---------:|---------:|--------:|----------:|
| CreateCalendar | 19,199.2 us | 269.51 us | 238.91 us | 750.0000 | 312.5000 | 62.5000 |  4,435 KB |
|   CreateButton |    127.1 us |   1.56 us |   1.46 us |   9.1553 |   1.2207 |       - |     43 KB |
|  CreateTextBox |    763.7 us |   5.69 us |   5.32 us |  46.8750 |   8.7891 |       - |    221 KB |

####  ResourceBenchmarks
Master:
|                  Method |      Mean |     Error |    StdDev | Allocated |
|------------------------ |----------:|----------:|----------:|----------:|
|         FindPreResource | 536.54 us |  5.687 us |  5.319 us |         - |
|        FindPostResource |  22.87 us |  0.209 us |  0.196 us |         - |
| FindNotExistingResource | 567.82 us | 10.430 us | 14.277 us |         - |

This PR:
|                  Method |      Mean |    Error |   StdDev | Allocated |
|------------------------ |----------:|---------:|---------:|----------:|
|         FindPreResource | 517.46 us | 5.802 us | 4.845 us |         - |
|        FindPostResource |  22.40 us | 0.360 us | 0.337 us |         - |
| FindNotExistingResource | 536.93 us | 4.226 us | 3.953 us |         - |

####  FluentBenchmark
Master:
|       Method |     Mean |   Error |  StdDev |   Gen 0 |  Gen 1 | Allocated |
|------------- |---------:|--------:|--------:|--------:|-------:|----------:|
| RepeatButton | 234.7 us | 4.67 us | 7.41 us | 14.1602 | 0.7324 |     65 KB |

This PR:
|       Method |     Mean |   Error |  StdDev |   Gen 0 |  Gen 1 | Allocated |
|------------- |---------:|--------:|--------:|--------:|-------:|----------:|
| RepeatButton | 188.2 us | 1.37 us | 1.22 us | 14.4043 | 0.7324 |     67 KB |

####  ThemeBenchmark
Master:
|           Method |             themeUri |      Mean |     Error |    StdDev |    Median | Allocated |
|----------------- |--------------------- |----------:|----------:|----------:|----------:|----------:|
| InitDefaultTheme | Dark |  6.803 ms | 0.1334 ms | 0.1248 ms |  6.758 ms |      1 MB |
| InitDefaultTheme | Light |  5.691 ms | 0.2675 ms | 0.7889 ms |  5.405 ms |      1 MB |
|  InitFluentTheme | Dark | 12.361 ms | 0.2471 ms | 0.7008 ms | 12.127 ms |      2 MB |
|  InitFluentTheme | Light | 12.091 ms | 0.2283 ms | 0.5337 ms | 12.035 ms |      2 MB |

This PR (only Dark):
|           Method |     Mean |     Error |    StdDev | Allocated |
|----------------- |---------:|----------:|----------:|----------:|
| InitDefaultTheme | 3.870 ms | 0.2114 ms | 0.6234 ms |      2 MB |
|  InitFluentTheme | 5.219 ms | 0.2442 ms | 0.7007 ms |      3 MB |

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0020612-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by grokys** [comment]
A few more benchmarks, this time `ControlsBenchmark`:

```
master

|         Method |        Mean |     Error |    StdDev |    Gen 0 |    Gen 1 |   Gen 2 | Allocated |
|--------------- |------------:|----------:|----------:|---------:|---------:|--------:|----------:|
| CreateCalendar | 21,481.0 us | 130.35 us | 121.93 us | 718.7500 | 312.5000 | 31.2500 |  4,275 KB |
|   CreateButton |    155.2 us |   0.75 us |   0.62 us |  10.0098 |   1.2207 |       - |     41 KB |
|  CreateTextBox |    903.5 us |   8.47 us |   7.92 us |  50.7813 |   9.7656 |       - |    218 KB |

themeresource

|         Method |        Mean |     Error |    StdDev |   Gen 0 |  Gen 1 | Allocated |
|--------------- |------------:|----------:|----------:|--------:|-------:|----------:|
| CreateCalendar | 17,545.7 us | 315.37 us | 705.36 us |       - |      - |  4,436 KB |
|   CreateButton |    149.7 us |   1.05 us |   0.98 us | 10.2539 | 1.2207 |     43 KB |
|  CreateTextBox |    864.6 us |   2.72 us |   2.54 us | 51.7578 | 9.7656 |    221 KB |
```

So there seems to be quite a decent speedup. More allocated memory, but that's offset by no GC.

**Comment by grokys** [comment]
I've not fully reviewed this PR, but so far I only have two comments:

- As mentioned in the issue, the term "theme" is overloaded. Maybe something like "theme variant" might be clearer?
- I'm a little concerned about dynamic resources now [having to make two subscriptions](https://github.com/AvaloniaUI/Avalonia/pull/8166/files#diff-bea90478b41c7180f29159ce6be8b3edc43ccab9d41882e30ffbb1bfe4eac924R143-R147) for each resource

I realise that this was done so that `DynamicResource` could be used, with no separate `ThemeResource`, but I wonder if this may hamper optimization in future?

**Comment by maxkatz6** [comment]
Internally we agreed on ThemeVariant property name as for this moment. Might be changed depending on ControlTheme PR status. We want to avoid naming conflicts.
ThemeControl should also be renamed to define only a scope for theme variant (dark/light...), and not the whole theme (fluent/default).
"ThemeDictionary" probably still makes sense as it is now.

**Comment by maxkatz6** [comment]
We now have Control.Theme property on each control.
So, we need some other naming options:
- `Control.ThemeVariant` (of `ThemeVariant` type) - after previous discussion this one was prefered
- `Control.ThemeMode` (of `ThemeMode` type) - sounds natural as for me
- `Control.ThemeScheme` (of `ThemeScheme` type)

Other frameworks:
- UWP - `ElementTheme` 
- UIKit (iOS) - `UIUserInterfaceStyle` (seems like community uses term `theme` there a lot)
- Flutter - `Brightness` (in a context of `Theme Brightness`)
- Electron - `Theme` (`nativeTheme`)
- HTML5 - `color-scheme` (`prefers-color-scheme`)

**Comment by grokys** [comment]
Yeah I think I still prefer `Control.ThemeVariant`. `colorTheme` from SwiftUI is a possibility, but given that it can control more than just colors I still prefer `ThemeVariant`. At least there's some precedent from GTK :)

**Comment by workgroupengineering** [comment]
AppThemeColor?

**Comment by maxkatz6** [comment]
so far:

master:
|                  Method |      Mean |    Error |   StdDev | Allocated |
|------------------------ |----------:|---------:|---------:|----------:|
|         FindPreResource | 586.43 us | 5.257 us | 4.917 us |         - |
|        FindPostResource |  22.92 us | 0.091 us | 0.081 us |         - |
| FindNotExistingResource | 567.80 us | 6.618 us | 6.190 us |         - |

this PR:
|                  Method |      Mean |     Error |    StdDev | Allocated |
|------------------------ |----------:|----------:|----------:|----------:|
|         FindPreResource | 632.33 us | 12.492 us | 23.768 us |         - |
|        FindPostResource |  27.60 us |  0.149 us |  0.132 us |         - |
| FindNotExistingResource | 627.10 us |  9.497 us |  7.930 us |         - |

master:
|       Method |     Mean |   Error |  StdDev |  Gen 0 |  Gen 1 | Allocated |
|------------- |---------:|--------:|--------:|-------:|-------:|----------:|
| RepeatButton | 148.3 us | 2.93 us | 3.91 us | 6.1035 | 2.1973 |     38 KB |

this PR:
|       Method |     Mean |   Error |  StdDev |  Gen 0 |  Gen 1 | Allocated |
|------------- |---------:|--------:|--------:|-------:|-------:|----------:|
| RepeatButton | 128.3 us | 2.46 us | 2.18 us | 6.3477 | 2.4414 |     39 KB |

master:
|          Method |  mode |     Mean |    Error |   StdDev |   Gen 0 |   Gen 1 | Allocated |
|---------------- |------ |---------:|---------:|---------:|--------:|--------:|----------:|
| InitFluentTheme | Light | 647.7 us |  8.80 us |  8.23 us | 84.9609 | 41.9922 |    525 KB |
| InitFluentTheme |  Dark | 629.2 us | 11.93 us | 11.71 us | 84.9609 | 41.9922 |    525 KB |
| InitSimpleTheme | Light | 314.9 us |  2.60 us |  2.30 us | 43.9453 | 14.6484 |    222 KB |
| InitSimpleTheme |  Dark | 310.1 us |  6.02 us |  5.91 us | 42.9688 | 15.6250 |    230 KB |

this PR:
|          Method |     Mean |   Error |   StdDev |    Gen 0 |   Gen 1 | Allocated |
|---------------- |---------:|--------:|---------:|---------:|--------:|----------:|
| InitFluentTheme | 949.5 us | 8.65 us |  8.10 us | 136.7188 | 68.3594 |    849 KB |
| InitSimpleTheme | 298.0 us | 5.89 us | 13.30 us |  46.8750 | 16.1133 |    229 KB |

**Comment by timunie** [comment]
> AppThemeColor?

For me that would conflict with AccentColor for newbies. Just my two cents.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0022956-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0023189-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by maxkatz6** [comment]
@DrWenz sorry to bother you, but I wonder how this PR affects startup time on your low end devices.
From my testing with control catalog and dotTrace on my machine (which isn't really a low end), I see something about 5-10% startup time degradation. With R2R there are no measurable difference at all.
Meanwhile micro benchmarks from above also show some degradation on resources lookup. But it's hardly reliable, as we don't have automated benchmarks for any big application as for now.

**Comment by timunie** [comment]
@maxkatz6 did you re-run your benchmark after your last edits? I can imagine that removing unnecessary foreach loops can improve performance unless it was never or only rarely hit.

**Comment by DrWenz** [comment]
> @DrWenz sorry to bother you, but I wonder how this PR affects startup time on your low end devices.
> From my testing with control catalog and dotTrace on my machine (which isn't really a low end), I see something about 5-10% startup time degradation. With R2R there are no measurable difference at all.
> Meanwhile micro benchmarks from above also show some degradation on resources lookup. But it's hardly reliable, as we don't have automated benchmarks for any big application as for now.

Sure I can do some tests, do you want a comparison with 0.10.18 or the last master?

**Comment by maxkatz6** [comment]
@DrWenz master. As we already know that master (and likely this PR) is faster than 0.18.0. 

@timunie yes, I did rerun some of the benchmarks

**Comment by DrWenz** [comment]
@maxkatz6 as I can see, FluentThemeMode is removed and Dark/Light is handled via separate ResourceDictionary. Any other changes I should consider or adjust in my app, in regards to this PR, when I run some tests?

**Comment by maxkatz6** [comment]
@DrWenz instead of FluentThemeMode there is an Application.ThemeVariant now. Which also can be set on each window independently. While having single resource dictionary for all of these variants. 

Existing custom resources will continue to work same way as before, so nothing else needs to be changed.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0023387-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0023672-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by maxkatz6** [comment]
To consider https://github.com/microsoft/WindowsAppSDK/issues/41 https://github.com/dotnet/winforms/issues/5166

**Comment by maxkatz6** [comment]
@AvaloniaUI/core and @amwx @SKProCH as maintainers of biggest third-party themes for Avalonia.
This PR is again ready to review, now with updated description which should cover everything (and basically be copy pasted into documentation page).
There are some issues with text rendering in Avalonia, unrelated to this PR, but makes harder to review. But at least please take a look at the API surface.
From what I checked, there shouldn't be any issues with integrating theme variants and theme dictionaries into third party themes. But just in case.

**Comment by amwx** [comment]
@maxkatz6 Can you try to fix the build so there's a package with the new API design? It'll be easier for me to test this than just look at the API surface.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0028798-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Comment by maxkatz6** [comment]
@amwx thanks, done.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0028801-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Comment by amwx** [comment]
@maxkatz6 Found one issue so far trying to integrate this with FluentAvalonia. Light and Dark switching works perfectly, but it will not work with other themes (HighContrast specifically). The resource lookups are coming back with transparent. What seems to be happening is (example):
Brush: `ApplicationPageBackgroundThemeBrush` 
Maps to: `SystemColorWindowColor`

When the DeferredItem factory runs for the Brush, the lookup for `SystemColorWindowColor` is done with `null` ThemeVariant which ends up looking in the Default dictionary (I've followed you so its Light), where that resource isn't defined so it ends up returning transparent. I think the Markup extensions are going to need context/theme awareness so the lookup is done correctly. 

Interestingly, when Dark (for example) mode is used, when that Brush initializes in `ResourceDictionary.TryGetValue()`, when the item factory runs, the very next call to `TryGetValue` is on the same dark theme resource dictionary. When using the HighContrast theme, the next call to `TryGetValue` after the factory isn't the theme dictionary, but the main ResourceDictionary on FluentAvaloniaTheme (which is of type `Styles`). Not really sure what's causing that difference.

Note: that I just kind of threw this PR into FA so it's possible I'm doing something that's messing with it. The only special thing I do is in FATheme I force search the Application ResourceDictionary first for overriding purposes. But I took that out to test and the same behavior occurs. I'll test some more tomorrow

Otherwise, I think the API itself is fine. IMO I'm not a huge fan of the "Default" variant (I don't like it in WinUI either) as it is ambiguous as to its meaning, especially if the meaning can be changed between Avalonia & third parties. I haven't looked closely enough to know if the ThemeVariant.Default member has a purpose, but at least in the ThemeDictionaries IMO I think we should use actual theme names and just implicitly map "Default" to "Light"

**Comment by timunie** [comment]
maybe `FallbackTheme` is a better naming than `Default`? At least for me it would be easier to understand what is meant by this, a fallback to search if the other lookup was not successful.

**Comment by maxkatz6** [comment]
The problem is, "ThemeVariant.Default" is used in two different ways depending on the context.
`RequestedTheme="Default"` will reset variant on the current control. Basically, sets it to "Unset".
`ResourceDictionary x:Key="Default"` will work as a fallback resource.

**Comment by SKProCH** [comment]
@maxkatz6 in Material.Avalonia we have `BaseThemeMode` (Light and Dark actually) and also `Primary` and `Secondary` colors (which may be any color). Do I understand correctly that in this case this change only covers `BaseThemeMode` in my case?

**Comment by maxkatz6** [comment]
@SKProCH yes, only dark/light mode, I don't think there is any meaningful way to abstract app accent colors across the app. As different themes might have one or many accent colors. Like material 3 on android which has three accent color.

**Comment by amwx** [comment]
@maxkatz6  Expanding on the issue I noted, take this repro into the Sandbox project:

MainWindow:
```Xaml
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x='http://schemas.microsoft.com/winfx/2006/xaml'
        x:Class="Sandbox.MainWindow"
        Background="#202020">

  <Border Width="300" Height="300"
          Background="{DynamicResource UniqueColorBrush}" />
  
</Window>
```

MainWindow.cs
```C#
public ThemeVariant CustomTheme = new ThemeVariant("Custom")
{
   InheritVariant = ThemeVariant.Light
};

public MainWindow()
{
    this.InitializeComponent();
    this.AttachDevTools();

    // A method to switch themes
    KeyDown += (s, e) =>
    {
        switch (e.Key)
        {
            case Avalonia.Input.Key.L:
                Application.Current.RequestedThemeVariant = ThemeVariant.Light;
                break;

            case Avalonia.Input.Key.D:
                Application.Current.RequestedThemeVariant = ThemeVariant.Dark;
                break;

            case Avalonia.Input.Key.Space:
                Application.Current.RequestedThemeVariant = CustomTheme;
                break;
        }
    };
}
```

Now, take this ResourceDictionary and put it in the Window's Resources:
```Xaml
<ResourceDictionary>
  <ResourceDictionary.ThemeDictionaries>
    <ResourceDictionary x:Key="Default">
      <Color x:Key="UniqueColor">Red</Color>

      <SolidColorBrush x:Key="UniqueColorBrush" Color="{StaticResource UniqueColor}" />
    </ResourceDictionary>
    <ResourceDictionary x:Key="Dark">
      <Color x:Key="UniqueColor">Blue</Color>

      <SolidColorBrush x:Key="UniqueColorBrush" Color="{StaticResource UniqueColor}" />
    </ResourceDictionary>
    <ResourceDictionary x:Key="Custom">
      <Color x:Key="UniqueColor">Purple</Color>

      <SolidColorBrush x:Key="UniqueColorBrush" Color="{StaticResource UniqueColor}" />
    </ResourceDictionary>
  </ResourceDictionary.ThemeDictionaries>
</ResourceDictionary>
```

So far this should all work as expected.

Now make the following changes:
- Remove the resources from the light/dark theme dictionaries, but leave the Color and Brush in the Custom dictionary 
- Change the binding on the brush color to a DynamicResource.

Switch to the Custom theme, then to either light or dark, then back to custom and the Border will be transparent. If you return the light/dark resources (keeping the custom theme as a DynamicResource) it will work. Interestingly, only the colors need to be kept in all dictionaries for the lookup to succeed, removing just the brushes from light/dark have no effect.

<br />

Now, remove the ResourceDictionary from the window. Copy the original (from above) to the Application resources. 

Change the binding on custom theme's brush to DynamicResource (but leave the others active). The lookup will fail and fall back to Light theme (Red) when you switch to the Custom theme (this worked in the Window's resources, but fails at the App level). Thus, not having the resources present in light/dark would lead to the lookup failing, and this is basically the same issue/result with the HighContrast theme I already noted.

It seems to be an issue with DynamicResource

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0029177-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Comment by maxkatz6** [comment]
@amwx I investigated this issue and found surprising bug with our bindings support. See https://github.com/AvaloniaUI/Avalonia/issues/10110

File: src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/AvaloniaXamlIlLanguageParseIntrinsics.cs [hunk 1]
```diff
@@ -291,6 +291,26 @@ public static bool TryConvert(AstTransformationContext context, IXamlAstValueNod
...
+                var foundConstProperty = types.ThemeVariant.Properties.FirstOrDefault(p =>
+                    p.Name == variantText && p.PropertyType == types.ThemeVariant);
+                var themeVariantTypeRef = new XamlAstClrTypeReference(node, types.ThemeVariant, false);
+                if (foundConstProperty is not null)
+                {
+                    result = new XamlStaticExtensionNode(new XamlAstObjectNode(node, node.Type), themeVariantTypeRef, foundConstProperty.Name);
+                    return true;
+                }
+
+                result = new XamlAstNewClrObjectNode(node, themeVariantTypeRef, types.ThemeVariantConstructor,
```
Comment (kekekeks) [inline src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/AvaloniaXamlIlLanguageParseIntrinsics.cs:306]: As discussed separately, auto-constructing an incomplete ThemeVariant instance is a bad idea

**Review by kekekeks — APPROVED** [review]
LGTM except for the compiler thingy

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0029334-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Comment by 0x90d** [comment]
Hi,

I just upgraded to Preview 5 and figured out the Fluent Mode property has been removed. It seems like I must set the mode in code using:

`Application.Current.RequestedThemeVariant = Avalonia.Styling.ThemeVariant.Dark;`

That seems to work. However I am also using `DataGrid` and the code above has no impact. The DataGrid is still themed in light mode. Can someone please quickly tell me what I am missing? List of [breaking changes](https://github.com/AvaloniaUI/Avalonia/wiki/Breaking-Changes#avaloniathemesfluent-and-simple) doesn't give me a clue

**Comment by maxkatz6** [comment]
@0x90d it's a bug, see https://github.com/AvaloniaUI/Avalonia/issues/9410

**Pull request merged**